monoio/src/io/async_read_rent.rs

```diff
@@ -43,7 +43,7 @@ pub trait AsyncReadRent {
 pub trait AsyncReadRentAt {
     /// Same as pread(2)
     fn read_at<T: IoBufMut>(
-        &mut self,
+        &self,
```

Comment:
It is useful to keep `&mut self`, refer to https://github.com/bytedance/monoio/pull/299#discussion_r1749526253

Reply:
It's not clear to me that `AsyncReadRent` and `AsyncReadRentAt` need to align their receiver types. I'm pretty sure all platforms that monoio targets can implement a non-`mut` read_at/write_at?

Reply:
It's not only targes IO but also middlewares needs to implement AsyncReadRentAt, such as if we want to implement AsyncReadRentAt for monoio type of https://doc.rust-lang.org/src/std/io/buffered/bufreader.rs.html#317-329, it needs inner mutations (fill buf)

Reply:
I agree with @ethe, there may be state(like buffer) inside the middlewares that implements AsyncReadRent. So for the struct itself, we can expose `&self` if possible; and for the trait we can keep `&mut self`.
Usually this style works well for stream(like tcp/uds) and file io, because even we expose `&self`, in most cases users want to make sure something like `write_all` not run in parallel(otherwise multiple `write` may results in data out of order).

Reply:
`AsyncReadRentAt` should only be implemented by types that support an efficient positional read operation, so shouldn't buffering occurring at a different level, e.g. [via a wrapper](https://github.com/compio-rs/compio/blob/master/compio-io/src/read/buf.rs#L25-L54)?

Reply:
The reason I'm pushing back a bit here is b/c a lot of things will sit downstream of this foundational trait. I'm currently exploring using it for writing a file format client, and needing to `Mutex`-wrap or `.clone()` file handles to allow parallel access for reading feels wasteful

Reply:
If we use buffered reader as a specific type, we would loose the polymorphism, we could not implement `AsyncReadRentAt` for buffered reader any more, I am a little bit worry about this. Considering the tradition: `Cursor` and `BufReader` in std, `impl Read for BufReader` allows user define a recursive `impl Read`.

Reply:
I think cloning fd takes few cost, what do you think about it @a10y ? I tend to keep file not syncable.

Reply:
Yea I can go ahead and update the PR to bring back the mutable receiver